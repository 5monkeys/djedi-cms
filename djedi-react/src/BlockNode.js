import React from "react";

export default function BlockNode() {
  return <span>BlockNode</span>;
}
