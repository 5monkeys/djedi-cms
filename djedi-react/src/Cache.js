export default class Cache {
  constructor({ ttl }) {
    this.ttl = ttl;
    this._cache = new Map();
  }

  get(key) {
    const item = this._cache.get(key);

    if (item == null) {
      return undefined;
    }

    const needsRefresh = Date.now() - item.timestamp > this.ttl;

    return { node: item.value, needsRefresh };
  }

  set(key, value) {
    this._cache.set(key, { timestamp: Date.now(), value });
  }

  delete(key) {
    this._cache.delete(key);
  }
}
