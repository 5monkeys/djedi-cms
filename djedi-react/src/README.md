<img alt="Djedi CMS" src="https://djedi-cms.org/_static/djedi-landscape.svg" width="500"/>

# djedi-react

[React] client for [Djedi CMS].

## Installation

```
npm install djedi-react
```

## Usage

Not yet.

## Development

* `npm start`: Start [Jest] in watch mode.
* `npm lint`: Run [ESLint] \(including [Prettier]).
* `npm lint:fix`: Autofix [ESLint] errors.
* `npm jest`: Run unit tests.
* `npm build`: Compile with [Babel].
* `npm test`: Check that everything works.
* `npm publish`: Publish to [npm], but only if `npm test` passes.

[babel]: http://babeljs.io/
[djedi cms]: http://djedi-cms.org/
[eslint]: https://eslint.org/
[jest]: https://facebook.github.io/
[npm]: https://www.npmjs.com/
[prettier]: https://prettier.io/
[react]: https://reactjs.org/
