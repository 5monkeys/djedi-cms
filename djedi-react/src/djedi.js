// This file does not contain anything React-specific. If need JS client for
// another framework this file could be extracted into its own package and be
// re-used.

import { applyUriDefaults, parseUri, stringifyUri } from "./uri";

/*
This class fetches and caches nodes, and keeps `window.DJEDI_NODES` up-todate.

Central interface:

    type Node = {
      // Input nodes: Possibly non-absolute URI.
      // Output nodes: Absolute, normalized URI.
      uri: string,
      // Input nodes: The default value, if any.
      // Output nodes: The final value, if any. The final value can be the
      // default value, or a new value entered by the user. In both cases the
      // backend might have rendered the value (such as markdown -> HTML).
      value: string | undefined,
    }

The admin sidebar expects the following mapping of all rendered nodes on the page:

    window.DJEDI_NODES = {
      "<uri>": "<default>",
    }
*/
export class Djedi {
  constructor() {
    this.options = makeDefaultOptions();

    this._nodes = new Map();
    this._renderedNodes = new Map();
    this._batch = makeEmptyBatch();
    this._DJEDI_NODES = {};

    if (typeof window !== "undefined") {
      if (window.DJEDI_NODES == null) {
        window.DJEDI_NODES = {};
      }
      this._DJEDI_NODES = window.DJEDI_NODES;
    }
  }

  get(passedNode, callback) {
    const node = this._normalizeNode(passedNode);
    const { uri } = node;
    const existingNode = this._nodes.get(uri);

    if (existingNode != null) {
      callback(existingNode);
      return;
    }

    this.loadMany([node]).then(
      results => {
        callback(results[uri] || new Error(`Missing result for node: ${uri}`));
      },
      error => {
        callback(error);
      }
    );
  }

  getBatched(passedNode, callback) {
    if (this.options.batchInterval <= 0) {
      this.get(passedNode, callback);
      return;
    }

    const node = this._normalizeNode(passedNode);
    const existingNode = this._nodes.get(node.uri);

    if (existingNode != null) {
      callback(existingNode);
      return;
    }

    const previous = this._batch.queue.get(node.uri) || {
      node,
      callbacks: [],
    };
    previous.callbacks.push(callback);

    if (this._batch.timeoutId != null) {
      return;
    }

    this._batch.timeoutId = setTimeout(
      this._flushBatch.bind(this),
      this.options.batchInterval
    );
  }

  async loadMany(nodes) {
    console.log("TODO: loadMany", nodes, this.options.baseUrl);
    try {
      const results = await Promise.resolve([]);
      this.addNodes(results);
      return results;
    } catch (passedError) {
      const error =
        passedError instanceof Error
          ? passedError
          : new Error(`Failed to load nodes: ${passedError}`);
      throw error;
    }
  }

  async loadByPrefix(prefixes) {
    const results = await Promise.resolve([]);
    console.log("TODO: preloadByPrefix", prefixes, results);
    this.addNodes(results);
    return results;
  }

  // Needed to pick up the results from `loadByPrefix` after server-side
  // rendering.
  addNodes(nodes) {
    for (const passedNode of nodes) {
      const node = this._normalizeNode(passedNode);
      this._nodes.set(node.uri, node);
    }
  }

  injectAdmin() {
    console.log("TODO: injectAdmin (maybe)");
  }

  reportRenderedNode(passedNode) {
    const node = this._normalizeNode(passedNode);
    const previous = this._renderedNodes.get(node.uri);

    if (previous != null) {
      if (previous.value !== node.value) {
        console.warn(
          "djedi-react: Rendering a node with with a different default value. That default will be ignored.",
          {
            uri: node.uri,
            prev: previous.value,
            next: node.value,
          }
        );
      }
      previous.numInstances++;
      return;
    }

    this._renderedNodes.set({
      value: node.value,
      numInstances: 1,
    });

    this._DJEDI_NODES[this._djediNodesUri(node.uri)] = node.value;
  }

  reportRemovedNode(passedUri) {
    const uri = this._normalizeUri(passedUri);
    const previous = this._renderedNodes.get(uri);

    if (previous == null) {
      return;
    }

    previous.numInstances--;

    if (previous.numInstances <= 0) {
      this._renderedNodes.delete(uri);
      delete this._DJEDI_NODES[this._djediNodesUri(uri)];
    }
  }

  resetOptions() {
    this.options = makeDefaultOptions();
  }

  resetNodes() {
    if (this._batch.timeoutId != null) {
      clearTimeout(this._batch.timeoutId);
    }

    this._nodes = new Map();
    this._renderedNodes = new Map();
    this._batch = makeEmptyBatch();
    this._DJEDI_NODES = {};

    if (typeof window !== "undefined") {
      window.DJEDI_NODES = this._DJEDI_NODES;
    }
  }

  element(uri) {
    const uriObject = this._parseUri(uri);
    return {
      tag: "span",
      attributes: {
        "data-i18n": this._stringifyUri({
          ...uriObject,
          scheme: "",
          ext: "",
          version: "",
        }),
      },
    };
  }

  _parseUri(uri) {
    const { defaults, namespaceByScheme, separators } = this.options.uri;
    const uriObject = parseUri(uri, separators);
    return applyUriDefaults(uriObject, defaults, namespaceByScheme);
  }

  _stringifyUri(uriObject) {
    return stringifyUri(uriObject, this.options.uri.separators);
  }

  _normalizeUri(uri) {
    return this._stringifyUri(this._parseUri(uri));
  }

  _normalizeNode(node) {
    return {
      ...node,
      uri: this._stringifyUri(this._parseUri(node.uri)),
    };
  }

  _djediNodesUri(uri) {
    const uriObject = this._parseUri(uri);
    return this._stringifyUri({ ...uriObject, version: "" });
  }

  async _flushBatch() {
    const { queue } = this._batch;
    const nodes = Array.from(queue).reduce((result, [uri, data]) => {
      result[uri] = data.node.value;
      return result;
    });

    this._batch = makeEmptyBatch();

    try {
      const results = this.loadMany(nodes);
      for (const [uri, data] of queue) {
        const value = results[uri];
        const node =
          value == null
            ? new Error(`Missing result for node: ${uri}`)
            : { uri, value };
        for (const callback of data.callbacks) {
          callback(node);
        }
      }
    } catch (error) {
      for (const [, data] of queue) {
        for (const callback of data.callbacks) {
          callback(error);
        }
      }
    }
  }
}

// This is a function, not a constant, since it can be mutated by the user.
function makeDefaultOptions() {
  return {
    baseUrl: "",
    batchInterval: 50, // ms
    defaultRender: {
      loading: "Loadingâ€¦",
      error: "Failed to fetch content ðŸ˜ž",
    },
    uri: {
      defaults: {
        scheme: "i18n",
        namespace: "",
        path: "",
        ext: "txt",
        version: "",
      },
      namespaceByScheme: {
        i18n: "en-us",
        l10n: "local",
        g11n: "global",
      },
      separators: {
        scheme: "://",
        namespace: "@",
        path: "/",
        ext: ".",
        version: "#",
      },
    },
  };
}

// This is a function, not a constant, since it will be mutated.
function makeEmptyBatch() {
  return {
    timeoutId: undefined,
    queue: new Map(),
  };
}

export default new Djedi();
