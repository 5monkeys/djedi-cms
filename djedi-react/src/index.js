export { default as djedi } from "./djedi";
export { default as Node, NodeContext } from "./Node";
export { default as md } from "dedent-js";
