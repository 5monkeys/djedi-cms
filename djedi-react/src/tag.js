export { default } from "dedent-js";
