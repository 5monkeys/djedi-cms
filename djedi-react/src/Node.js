import PropTypes from "prop-types";
import React from "react";
import dedent from "dedent";

import djedi from "./djedi";

const POTENTIAL_HTML = /[<&]/;

const propTypes = {
  uri: PropTypes.string.isRequired,
  children: PropTypes.string, // Yes, a *string*!
  edit: PropTypes.bool,
  render: PropTypes.func,
  // ...kwargs: {[key: string]: string}.
};

const defaultProps = {
  edit: true,
  children: undefined,
  render: undefined,
};

export default class Node extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      node: undefined,
    };

    this.mounted = false;

    // Must be done in the constructor rather than `componentDidMount` because
    // of server-side rendering.
    this._get();
  }

  componentDidMount() {
    this.mounted = true;
    djedi.reportRenderedNode({
      uri: this.props.uri,
      value: this._getDefault(),
    });
  }

  componentDidUpdate(prevProps) {
    if (this.props.uri !== prevProps.uri) {
      this._get();
      djedi.reportRemovedNode(prevProps.uri);
      djedi.reportRenderedNode({
        uri: this.props.uri,
        value: this._getDefault(),
      });
    }

    if (
      this.props.uri === prevProps.uri &&
      this.props.children !== prevProps.children
    ) {
      console.error(
        "djedi-react: Changing the default value of a node is not supported.",
        { prev: prevProps.children, next: this.props.children }
      );
    }
  }

  componentWillUnmount() {
    this.mounted = false;
    djedi.reportRemovedNode(this.props.uri);
  }

  _get() {
    djedi.getBatched(
      { uri: this.props.uri, value: this._getDefault() },
      node => {
        if (this.mounted) {
          this.setState({ node });
        } else {
          // If the node already exists in cache, this callback is called
          // immediately (synchronously). If this happens in the `_get()` call in
          // `constructor`, where it is not allowed to call `setState`, mutate
          // `this.state` directly instead (which is fine in the `constructor`).
          // eslint-disable-next-line react/no-direct-mutation-state
          this.state.node = node;
        }
      }
    );
  }

  _getDefault() {
    const { children } = this.props;
    return children == null ? undefined : dedent(String(children));
  }

  render() {
    const {
      uri,
      children,
      edit,
      // Using a destructuring default rather than `defaultProps` for `render`,
      // since `djedi.options` may change.
      render = djedi.options.defaultRender,
      // Make sure to destructure all props above (even ones unused in this
      // method) so that `kwargs` only contains non-props.
      ...kwargs
    } = this.props;
    const { node } = this.state;

    if (node == null) {
      return render({ type: "loading" });
    }

    if (node instanceof Error) {
      return render({ type: "error", error: node });
    }

    const value = interpolate(node.value || "", kwargs);
    const element = djedi.element(node.uri);

    // This is mostly to make the test snapshots easier to read. Might be faster
    // not using `dangerouslySetInnerHTML` unnecessarily, too.
    const hasHtml = POTENTIAL_HTML.test(value);

    const content = edit ? (
      <element.tag
        {...element.attributes}
        dangerouslySetInnerHTML={hasHtml ? { __html: value } : undefined}
      >
        {hasHtml ? undefined : value}
      </element.tag>
    ) : (
      // WARNING: It’s not possible to render HTML without a wrapper element in
      // React. So you can’t do `edit={false}` and expect the node value to be
      // treated as HTML.
      value
    );

    return render({ type: "success", content });
  }
}

Node.propTypes = propTypes;
Node.defaultProps = defaultProps;

const INNER = /[^{}[\]\s]+/.source;
const INTERPOLATION_REGEX = RegExp(`\\{${INNER}\\}|\\[${INNER}\\]`, "g");

/*
This intentionally only supports `{key}`, not any of the fancy Python string
formatting extras. If `key` maps to a string in `kwargs`, that string is
inserted. Otherwise it is kept as-is. No escaping, no nothing. KISS and
YAGNI. If somebody does want fancy extras, they can do it themselves before
putting the value in `kwargs`.

Also support `[key]`, since `{foo}` is already used in JSX syntax:

    const user = "Bob"
    <Node uri="test">Hello, {user}!</Node>
    <Node uri="test">Hello, [user]!</Node>
    <Node uri="test">{`Hello, {user}!`}</Node>
*/
function interpolate(string, kwargs) {
  return string.replace(INTERPOLATION_REGEX, match => {
    const key = match.slice(1, -1);
    return Object.prototype.hasOwnProperty.call(kwargs, key)
      ? kwargs[key]
      : match;
  });
}
