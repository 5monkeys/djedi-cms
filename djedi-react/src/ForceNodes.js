import PropTypes from "prop-types";
import React from "react";

import Node from "./Node";

ForceNodes.propTypes = {
  children: PropTypes.node.isRequired,
};

export default function ForceNodes({ children }) {
  return React.Children.map(children, child => {
    if (child == null || child.type !== Node) {
      throw new TypeError(
        "djedi-react: <ForceNodes> only accepts <Node>s as children."
      );
    }
    return React.cloneElement(child, { render: () => null });
  });
}
