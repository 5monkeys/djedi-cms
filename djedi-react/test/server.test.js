/**
 * @jest-environment node
 */

import React from "react";
import dedent from "dedent";
import renderer from "react-test-renderer";

import { Node, djedi, md } from "../src";

import { fetch, resetAll, simpleNodeResponse, wait } from "./helpers";

jest.useFakeTimers();

beforeEach(() => {
  resetAll();
});

test("it renders loading and then the node", async () => {
  fetch(
    simpleNodeResponse(
      "test",
      dedent`
        <h1>Heading</h1>
        <p>And a <a href="{url}">link</a>.</p>
      `
    )
  );
  const component = renderer.create(
    <Node uri="test" url="https://example.com/">{md`
      # Heading

      And a [link]({url}).
    `}</Node>
  );
  djedi.options.baseUrl = "http://internal:3210";
  expect(component.toJSON()).toMatchSnapshot("loading");
  await wait();
  expect(fetch.mockFn.mock.calls).toMatchSnapshot("api call");
  expect(component.toJSON()).toMatchSnapshot("with value");
  expect(typeof window).toBe("undefined");
});

test("djedi.injectAdmin does not do anything", async () => {
  const injected = await djedi.injectAdmin();
  expect(injected).toBe(false);
  expect(fetch.mockFn).toHaveBeenCalledTimes(0);
  expect(typeof document).toBe("undefined");
});

test("when rendering the same view twice, djedi.prefetch results in the same nodes", async () => {
  fetch({
    ...simpleNodeResponse("1", "1"),
    ...simpleNodeResponse("2", "2"),
  });

  function Page() {
    return (
      <div>
        <Node uri="1" />
        <Node uri="2" />
      </div>
    );
  }

  djedi.reportPrefetchableNode({ uri: "1", value: undefined });
  djedi.reportPrefetchableNode({ uri: "2", value: undefined });

  const nodes1 = await djedi.prefetch();
  expect(fetch.mockFn).toHaveBeenCalledTimes(1);
  const component1 = renderer.create(<Page />);
  const tree1 = component1.toJSON();

  await wait();

  const nodes2 = await djedi.prefetch();
  expect(fetch.mockFn).toHaveBeenCalledTimes(1);
  const component2 = renderer.create(<Page />);
  const tree2 = component2.toJSON();

  expect(nodes2).toEqual(nodes1);
  expect(tree2).toEqual(tree1);
});
