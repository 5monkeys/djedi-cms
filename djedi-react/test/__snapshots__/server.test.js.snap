// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders loading and then the node: api call 1`] = `
Array [
  Array [
    "http://internal:3210/djedi/load_many",
    Object {
      "body": Object {
        "i18n://en-us@test.txt": "# Heading

And a [link]({url}).",
      },
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`it renders loading and then the node: loading 1`] = `"Loadingâ€¦"`;

exports[`it renders loading and then the node: with value 1`] = `
<span
  dangerouslySetInnerHTML={
    Object {
      "__html": "<h1>Heading</h1>
<p>And a <a href=\\"http://example.com/\\">link</a>.</p>",
    }
  }
  data-i18n="en-us@test"
/>
`;
